unity/src/main/java/io/delta/unity/UCCatalogManagedCommitter.java

```diff
@@ -103,4 +117,103 @@ private void validateLogPathBelongsToThisUcTable(CommitMetadata cm) {
         expectedDeltaLogPathNormalized,
         providedDeltaLogPathNormalized);
   }
+
+  private FileStatus writeStagedCommitFile(
+      Engine engine, CloseableIterator<Row> finalizedActions, CommitMetadata commitMetadata)
+      throws CommitFailedException {
+    final String stagedCommitFilePath = commitMetadata.getNewStagedCommitFilePath();
+
+    return timeCheckedOperation(
+        logger,
+        "Write staged commit file: " + stagedCommitFilePath,
+        ucTableId,
+        () -> {
+          try {
+            // Do not use Put-If-Absent for staged commit files since we assume that UUID-based
+            // commit files are globally unique, and so we will never have concurrent writers
+            // attempting to write the same commit file.
+
+            engine
+                .getJsonHandler()
+                .writeJsonFileAtomically(
+                    stagedCommitFilePath, finalizedActions, true /* overwrite */);
+
+            // TODO: [delta-io/delta#5021] Use FileSystemClient::getFileStatus API instead
+            return FileStatus.of(stagedCommitFilePath);
+          } catch (IOException ex) {
+            throw new CommitFailedException(
+                RETRYABLE,
+                NOT_CONFLICT,
+                "Failed to write staged commit file due to: " + ex.getMessage(),
+                ex);
+          }
+        });
+  }
+
+  private void commitToUC(CommitMetadata commitMetadata, FileStatus kernelStagedCommitFileStatus)
+      throws CommitFailedException {
+    timeCheckedOperation(
+        logger,
+        "Commit staged commit file to UC: " + kernelStagedCommitFileStatus.getPath(),
+        ucTableId,
+        () -> {
+          // commitToUc is only for normal catalog WRITES, not for CREATE, or UPGRADE, or DOWNGRADE,
+          // or anything filesystem related.
+          checkState(
+              commitMetadata.getCommitType() == CommitMetadata.CommitType.CATALOG_WRITE,
+              "Only supported commit type is CATALOG_WRITE, but got: %s");
+
+          try {
+            ucClient.commit(
+                ucTableId,
+                tablePath.toUri(),
+                Optional.of(getUCCommitPayload(commitMetadata, kernelStagedCommitFileStatus)),
+                Optional.empty() /* lastKnownBackfilledVersion */,
+                false /* isDisown */,
+                Optional.empty() /* newMetadata */, // TODO: support sending newProtocol
+                Optional.empty()); // TODO: support sending newProtocol
+            return null;
+          } catch (io.delta.storage.commit.CommitFailedException cfe) {
+            throw storageCFEtoKernelCFE(cfe);
```

Comment:
is this the error we'd hit if it is a conflict?

Reply:
Yup!